lib/ukallocstack/stack.c

```diff
@@ -0,0 +1,277 @@
+#endif /* CONFIG_LIBUKVMEM */
+};
+
+static struct uk_allocstack *to_allocstack(struct uk_alloc *a)
+{
+	UK_ASSERT(a);
+
+	return __containerof(a, struct uk_allocstack, a);
+}
+
+static void *stack_malloc(struct uk_alloc *a, __sz size)
+{
+	__vaddr_t sp __maybe_unused;
+	struct uk_allocstack *sa;
+	int rc __maybe_unused;
+	__vaddr_t vaddr;
+	__sz real_size;
+
+	UK_ASSERT(a);
+
+	if (!size)
+		return NULL;
+
+	sa = to_allocstack(a);
+
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	real_size = PAGE_ALIGN_UP(size) + PAGE_SIZE;
+	vaddr = __VADDR_ANY;
+	rc = uk_vma_map_stack(sa->vas,
+			      &vaddr,
+			      real_size,
+			      UK_VMA_MAP_UNINITIALIZED,
+			      NULL,
+			      sa->initial_size);
+	if (unlikely(rc)) {
+		uk_pr_err("Failed to create stack VMA of size 0x%lx\n",
+			  real_size);
+		return NULL;
+	}
+#else /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+	real_size = size;
+	vaddr = (__vaddr_t)uk_memalign(sa->og_a, UKARCH_SP_ALIGN, real_size);
+	if (unlikely(!vaddr)) {
+		uk_pr_err("Failed to allocate stack of sizegth 0x%lx from "
+			  "allocator %p\n", real_size, sa->og_a);
+		return NULL;
+	}
+
+#if CONFIG_LIBUKVMEM
+	sp = ukarch_gen_sp(vaddr, real_size);
+	rc = uk_vma_advise(sa->vas,
+			   PAGE_ALIGN_DOWN((__vaddr_t)
+					   (PAGE_ALIGN_UP(sp) -
+					    PAGE_ALIGN_UP(sa->initial_size))),
+			   PAGE_ALIGN_UP(sa->initial_size),
+			   UK_VMA_ADV_WILLNEED,
+			   UK_VMA_FLAG_UNINITIALIZED);
+	if (unlikely(rc)) {
+		uk_pr_err("Could not pre-map stack initial size\n");
+		return NULL;
+	}
+#endif /* CONFIG_LIBUKVMEM */
+#endif /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+
+	return (void *)vaddr;
+}
+
+static void stack_free(struct uk_alloc *a, void *ptr)
+{
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	const struct uk_vma *vma __maybe_unused;
+#endif /* CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+	__vaddr_t vaddr __maybe_unused;
+	struct uk_allocstack *sa;
+	int rc __maybe_unused;
+
+	UK_ASSERT(a);
+
+	if (!ptr)
+		return;
+
+	sa = to_allocstack(a);
+
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	vaddr = (__vaddr_t)ptr;
+	vma = uk_vma_find(sa->vas, vaddr);
+	/* Crash if pointer given to free is not part of any VMA. This would
+	 * normally lead to a heap corruption.
+	 */
+	UK_ASSERT(vma);
+
+	rc = uk_vma_unmap(sa->vas,
+			  vaddr,
+			  vma->end - vma->start,
+			  0);
+	UK_ASSERT(!rc);
+#else /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+	uk_free(sa->og_a, ptr);
+#endif /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+}
+
+static void *stack_memalign(struct uk_alloc *a, __sz align, __sz size)
+{
+	__vaddr_t sp __maybe_unused;
+	struct uk_allocstack *sa;
+	int rc __maybe_unused;
+	__vaddr_t vaddr;
+	__sz real_size;
+
+	UK_ASSERT(a);
+	UK_ASSERT(align >= UKARCH_SP_ALIGN);
+	UK_ASSERT(IS_ALIGNED(align,  UKARCH_SP_ALIGN));
+
+	if (!size)
+		return NULL;
+
+	sa = to_allocstack(a);
+
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	UK_ASSERT(align <= PAGE_SIZE);
+
+	real_size = PAGE_ALIGN_UP(size) + PAGE_SIZE;
+	vaddr = __VADDR_ANY;
+	rc = uk_vma_map_stack(sa->vas,
```

Comment:
Shouldn't you adjust the return vaddr to the requested alignment?

Reply:
If using page guards then we allocate stack VMA's instead. So, the return address will always be page-aligned and the maximum allowed alignment is PAGE_SIZE and smaller assignments must be a divisor of PAGE_SIZE.

I wil add a comment

Reply:
Okay since we have a fundamental limitation on alignment by guards, perhaps we should at least document this constraint, given that it deviates from the expected behavior, and also add a check.